File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/RSASignedDataInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import com.jetbrains.signatureverifier.serialization.toDLSequence
import com.jetbrains.signatureverifier.serialization.toPrimitiveDLSet
import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.*
import org.bouncycastle.asn1.cms.ContentInfo
import org.bouncycastle.asn1.x509.AlgorithmIdentifier

@Serializable
data class RSASignedDataInfo(
  val identifier: TextualInfo,
  val explicit: Boolean,
  val tagNo: Int,
  val version: TextualInfo,
  val digestAlgorithmsInfo: List<AlgorithmInfo>,
  val encapContentInfo: EncapContentInfo,
  val certificates: TaggedObjectInfo, //List<CertificateInfo>,
//  val crls: List<EncodableInfo>?,
  val counterSignatureInfos: List<CounterSignatureInfo>
) : EncodableInfo {
  companion object {
    fun getInstance(originalSequence: ASN1Sequence): RSASignedDataInfo {
      val identifier = TextualInfo.getInstance(originalSequence.first())

      val tagged = originalSequence.getObjectAt(1)
      val tagNo = (tagged as ASN1TaggedObject).tagNo
      val explicit = tagged.isExplicit

      val sequence = tagged.baseObject as DLSequence

      val iterator = sequence.iterator()
      val version = TextualInfo.getInstance(iterator.next())
      val algorithms = (iterator.next() as DLSet).map { AlgorithmInfo(AlgorithmIdentifier.getInstance(it)) }
      val encapContentInfo = EncapContentInfo.getInstance(
        ContentInfo.getInstance(iterator.next())
      )

      val certificates = iterator.next().let {
        TaggedObjectInfo(
          (it as DLTaggedObject).isExplicit,
          it.tagNo,
          SequenceInfo(
            (it.baseObject as DLSequence).map { obj ->
              CertificateInfo.getInstance(obj.toASN1Primitive())
            }
          )
        )
      }

      val counterSignatures = (iterator.next() as DLSet).map {
        CounterSignatureInfo.getInstance(it as DLSequence)
      }

      return RSASignedDataInfo(
        identifier,
        explicit,
        tagNo,
        version,
        algorithms,
        encapContentInfo,
        certificates,
        counterSignatures
      )
    }
  }

  override fun toPrimitive(): ASN1Primitive = listOf(
    identifier.toPrimitive(),
    TaggedObjectInfo.getTaggedObject(
      explicit,
      tagNo,
      listOf(
        version.toPrimitive(),
        digestAlgorithmsInfo.toPrimitiveDLSet(),
        encapContentInfo.toPrimitive(),
        certificates.toPrimitive(),
        counterSignatureInfos.toPrimitiveDLSet()
      ).toDLSequence()
    )
  ).toDLSequence()

}

File: net/JetBrains.Serialization/src/Dataholders/RSASignedDataInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Asn1.Cms;
using Org.BouncyCastle.Asn1.X509;

namespace JetBrains.Serialization;

[JsonObject(MemberSerialization.Fields)]
public class RSASignedDataInfo : IEncodableInfo
{
  [JsonProperty("Identifier")] private TextualInfo _identifier;
  [JsonProperty("Explicit")] private bool _explicit;
  [JsonProperty("TagNo")] private int _tagNo;
  [JsonProperty("Version")] private TextualInfo _version;
  [JsonProperty("DigestAlgorithmsInfo")] private List<AlgorithmInfo> _digestAlgorithmsInfo;
  [JsonProperty("EncapContentInfo")] private EncapContentInfo _encapContentInfo;
  [JsonProperty("Certificates")] private TaggedObjectInfo _certificates;

  [JsonProperty("CounterSignatureInfos")]
  private List<CounterSignatureInfo> _counterSignatureInfos;

  [JsonConstructor]
  public RSASignedDataInfo(
    TextualInfo identifier,
    bool isExplicit,
    int tagNo,
    TextualInfo version,
    List<AlgorithmInfo> digestAlgorithmsInfo,
    EncapContentInfo encapContentInfo,
    TaggedObjectInfo certificates,
    List<CounterSignatureInfo> counterSignatureInfos)
  {
    _identifier = identifier;
    _explicit = isExplicit;
    _tagNo = tagNo;
    _version = version;
    _digestAlgorithmsInfo = digestAlgorithmsInfo;
    _encapContentInfo = encapContentInfo;
    _certificates = certificates;
    _counterSignatureInfos = counterSignatureInfos;
  }

  public static RSASignedDataInfo GetInstance(Asn1Sequence originalSequence)
  {
    var identifier = TextualInfo.GetInstance(originalSequence[0]);

    var tagged = originalSequence[1] as DerTaggedObject;
    int tagNo = tagged!.TagNo;

    var sequence = tagged.GetObject() as DerSequence;

    var iterator = sequence!.GetEnumerator();
    iterator.MoveNext();

    var version = TextualInfo.GetInstance((Asn1Encodable)iterator.Current!);
    iterator.MoveNext();

    var algorithms = ((DerSet)iterator.Current!)
      .ToArray()
      .Select(it => new AlgorithmInfo(AlgorithmIdentifier.GetInstance(it))).ToList();
    iterator.MoveNext();

    var encapContentInfo = EncapContentInfo.GetInstance(ContentInfo.GetInstance(iterator.Current));
    iterator.MoveNext();

    var certificateObject = (DerTaggedObject)iterator.Current;
    var certificates = new TaggedObjectInfo(
      certificateObject.IsExplicit(),
      certificateObject.TagNo,
      new SequenceInfo(((DerSequence)certificateObject.GetObject()).ToArray()
        .Select(it => CertificateInfo.GetInstance(it.ToAsn1Object())).ToList())
    );
    iterator.MoveNext();

    var counterSignaturesSet = (DerSet)iterator.Current;

    var counterSignatures =
      counterSignaturesSet.ToArray()
        .Select(it => CounterSignatureInfo.GetInstance((DerSequence)it)).ToList();


    return new RSASignedDataInfo(
      identifier,
      tagged.IsExplicit(),
      tagNo,
      version,
      algorithms,
      encapContentInfo,
      certificates,
      counterSignatures
    );
  }

  public Asn1Encodable ToPrimitive() => new List<Asn1Encodable?>
    {
      _identifier.ToPrimitive(),
      TaggedObjectInfo.GetTaggedObject(
        _explicit,
        _tagNo,
        new List<Asn1Encodable?>
        {
          _version.ToPrimitive(),
          _digestAlgorithmsInfo.ToPrimitiveDerSet(),
          _encapContentInfo.ToPrimitive(),
          _certificates.ToPrimitive(),
          _counterSignatureInfos.ToPrimitiveDerSet()
        }.ToDerSequence()
      )
    }
    .ToDerSequence();
}